packages/core/lib/callback-runner.js

```diff
@@ -0,0 +1,46 @@
+const some = require('./async-every')
+
+module.exports = (callbacks, event, onCallbackError, cb) => {
+  // This function is how we support different kinds of callback:
+  //  - synchronous - return value
+  //  - node-style async with callback - cb(err, value)
+  //  - promise/thenable - resolve(value)
+  // It normalises each of these into the lowest common denominator – a node-style callback
+  const runMaybeAsyncCallback = (fn, cb) => {
+    if (typeof fn !== 'function') return cb(null, true)
```

Comment:
I wonder if this also ought to be async i.e. called within `setTimeout`? Also applies to the other sync `cb` calls. As it stands `runMaybeAsyncCallback` is maybe going to call `cb` sync or async. Would it be cleaner if it was always async?

Not sure it matters in our case but it can have an effect if consumers are expecting something to be called on the next event loop but it gets called synchronously.

Reply:
I totally see your point. Made them all async now.

Reply:
Reneged on this (pushed a commit and then removed it) – there are whole bunch of unit tests that fail because they run synchronously after calling `notify()` – expecting the synchronous callbacks to have run. This isn't a problem in and of itself, but it highlighted that a lot of the internal callbacks are synchronous and they are used to get the device/app state at the moment of an error – if they run asynchronously, there's the chance for things to change after the error has happened.

One of the notifier spec requirements is that "callbacks run in the context of the error" and if we enforce that even sync callbacks should be run asynchronously I'd say it violates that.